[MM-21494] autocomplete suggestions

iomodo opened the pull request:
#### Summary
This PR will add support of the autocomplete suggestions

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-21494

#### Related PRs
server - https://github.com/mattermost/mattermost-server/pull/14259
redux - https://github.com/mattermost/mattermost-redux/pull/1121
demo-plugin - https://github.com/mattermost/mattermost-plugin-demo/pull/99

iomodo commented:
/update-branch

mickmister commented on components/suggestion/command_provider.jsx:
```diff
@@ -46,35 +46,25 @@ export class CommandSuggestion extends Suggestion {
-                            return;
-                        }
-
-                        if (cmd.trigger !== 'shortcuts' || !UserAgent.isMobile()) {
```
Is the `shortcuts` command filtering done server-side now?

levb commented:
Merge issues, @iomodo?

hanzei commented:
I kind of screw up up the branching. By merging `master` into this branch merging master  into `MM-20684-Slash-Command-Autocomplete` is also needed. Let me know if I should take care of this @iomodo.

hanzei commented:
@levb https://github.com/mattermost/mattermost-webapp/pull/5289/commits/a35f1ec7c6020043669bbcfdc13449b6b19a80e1 is the only "real" change in this PR. The rest is just `master` commits.

levb commented:
@hanzei @iomodo the commit looks good to me; but the PR needs to be either fixed or re-cut.

iomodo commented:
Addressed an issue raised by @mickmister 
Changed UX a bit as suggested by @asaadmahmood 
Reverted changes for mobile as mobile changes will need working closely with mobile team.

levb commented:
I am not qualified to review this, plus I'd like to involve more people in reviewing the overall autocomplete implementation. Hopefully @crspeller has the cycles to review?

iomodo commented:
Added support of the icons.
![autocomplete](https://user-images.githubusercontent.com/2340127/81312091-43ca1500-9097-11ea-903e-1c9891945f9e.gif)

mickmister commented on components/suggestion/command_provider.jsx:
```diff
@@ -93,4 +112,39 @@ export default class CommandProvider extends Provider {
     }
+
+    handleWebapp(pretext, resultCallback) {
+        if (pretext.startsWith('/')) {
```
@iomodo do you mind changing this to return early on a non slash command instead of having most of the function indented? Same with `handleMobile`
```suggestion
        if (!pretext.startsWith('/')) {
        	return false;
        }
```
